File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/domain/interactor/bike/UpdateBikeMetaDatUseCase.java
package com.lattis.ellipse.domain.interactor.bike;

import com.lattis.ellipse.domain.executor.PostExecutionThread;
import com.lattis.ellipse.domain.executor.ThreadExecutor;
import com.lattis.ellipse.domain.interactor.UseCase;
import com.lattis.ellipse.domain.repository.BikeRepository;

import javax.inject.Inject;

import io.reactivex.Observable;

/**
 * Created by ssd3 on 8/31/17.
 */

public class UpdateBikeMetaDatUseCase extends UseCase<Boolean> {

    private BikeRepository bikeRepository;
    private String firmware_version;
    private Boolean shackle_jam;
    private int bike_battery_level = -1;
    private int lock_battery_level = -1;
    private int bike_id;


    @Inject
    public UpdateBikeMetaDatUseCase(ThreadExecutor threadExecutor,
                                    PostExecutionThread postExecutionThread,
                                    BikeRepository bikeRepository) {
        super(threadExecutor, postExecutionThread);
        this.bikeRepository = bikeRepository;
    }


    public UpdateBikeMetaDatUseCase withFirmWare(String firmware_version) {
        this.firmware_version = firmware_version;
        return this;
    }

    public UpdateBikeMetaDatUseCase withShackleJamStatus(Boolean shackle_jam) {
        this.shackle_jam = shackle_jam;
        return this;
    }

    public UpdateBikeMetaDatUseCase withLockBattery(Integer lock_battery_level) {
        if(lock_battery_level!=null) {
            this.lock_battery_level = lock_battery_level;
        }
        return this;
    }

    public UpdateBikeMetaDatUseCase withBikeBattery(int bike_battery_level) {
        this.bike_battery_level = bike_battery_level;
        return this;
    }

    public UpdateBikeMetaDatUseCase forBike(int bike_id) {
        this.bike_id = bike_id;
        return this;
    }



    @Override
    protected Observable<Boolean> buildUseCaseObservable() {
        return bikeRepository.updateBikeMetaData(bike_id, bike_battery_level, lock_battery_level, firmware_version, shackle_jam);
    }

}

File: velolabs_repos/lattis_android_v2/domain/src/main/java/com/lattis/domain/usecase/bike/UpdateBikeMetaDatUseCase.kt
package com.lattis.domain.usecase.bike

import com.lattis.domain.executor.PostExecutionThread
import com.lattis.domain.executor.ThreadExecutor
import com.lattis.domain.repository.BikeRepository
import com.lattis.domain.usecase.base.UseCase
import io.reactivex.rxjava3.core.Observable
import javax.inject.Inject

class UpdateBikeMetaDatUseCase @Inject constructor(
    threadExecutor: ThreadExecutor,
    postExecutionThread: PostExecutionThread,
    private val bikeRepository: BikeRepository
) : UseCase<Boolean>(threadExecutor, postExecutionThread) {

    private var firmware_version: String? = null
    private var shackle_jam: Boolean? = null
    private var bike_battery_level = -1
    private var lock_battery_level = -1
    private var bike_id = 0

    fun withFirmWare(firmware_version: String?): UpdateBikeMetaDatUseCase {
        this.firmware_version = firmware_version
        return this
    }

    fun withShackleJamStatus(shackle_jam: Boolean?): UpdateBikeMetaDatUseCase {
        this.shackle_jam = shackle_jam
        return this
    }

    fun withLockBattery(lock_battery_level: Int?): UpdateBikeMetaDatUseCase {
        if (lock_battery_level != null) {
            this.lock_battery_level = lock_battery_level
        }
        return this
    }

    fun withBikeBattery(bike_battery_level: Int): UpdateBikeMetaDatUseCase {
        this.bike_battery_level = bike_battery_level
        return this
    }

    fun forBike(bike_id: Int): UpdateBikeMetaDatUseCase {
        this.bike_id = bike_id
        return this
    }

    override fun buildUseCaseObservable(): Observable<Boolean> {
        return bikeRepository.updateBikeMetaData(
            bike_id,
            bike_battery_level,
            lock_battery_level,
            firmware_version,
            shackle_jam!!
        )
    }

}